Demos/Eyes/EyesWidget.cpp

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright (c) 2020, Sergey Bugaev <bugaevc@serenityos.org>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "EyesWidget.h"
+#include <LibGUI/Painter.h>
+#include <LibGUI/Window.h>
+#include <LibGUI/WindowServerConnection.h>
+#include <LibGfx/Palette.h>
+#include <math.h>
+
+EyesWidget::~EyesWidget()
+{
+}
+
+void EyesWidget::track_cursor_globally()
+{
+    ASSERT(window());
+    auto window_id = window()->window_id();
+    ASSERT(window_id >= 0);
+
+    set_global_cursor_tracking(true);
+    GUI::WindowServerConnection::the().send_sync<Messages::WindowServer::SetGlobalCursorTracking>(window_id, true);
+}
+
+void EyesWidget::mousemove_event(GUI::MouseEvent& event)
+{
+    m_mouse_position = event.position();
+    update();
+}
+
+void EyesWidget::paint_event(GUI::PaintEvent& event)
+{
+    GUI::Painter painter(*this);
+
+    painter.clear_rect(event.rect(), Gfx::Color());
+
+    for (int i = 0; i < m_num_eyes; i++)
+        render_eyeball(i, painter);
+}
+
+void EyesWidget::render_eyeball(int index, GUI::Painter& painter)
+{
+    auto eye_width = width() / m_num_eyes;
+    Gfx::Rect bounds { index * eye_width, 0, eye_width, height() };
+    auto thickness = max((width() + height()) / 15, 1);
+
+    bounds.shrink(width() / 25, 0);
+    painter.fill_ellipse(bounds, palette().base_text());
+    bounds.shrink(thickness, thickness);
+    painter.fill_ellipse(bounds, palette().base());
+
+    // OK, here comes a bunch of write-only math.
+    // You are not expected to understand this.
```

Comment:
...doesn't mean it has to look like a data scientist wrote it...can't we have more descriptive names than `ksq`? even `k_squared` would look better imho...

Reply:
Honestly, `ksq` is the least of with the problems with it. I have no idea how to make math like this readable.

Reply:
OK, let me try.

Reply:
well the math part of it would have to stay, but at the very least, the names could say what you want, I assume "da" is "distance_to_a" or something?

Reply:
Worse, _a_ was radius in one of the directions, and Dₐ was the corresponding diameter (twice the radius). Hold on, I'm trying to translate this to Serenity-style naming, and fix a division by zero.

Reply:
:facepalm: 
only robotics people call radii _a_, that's so silly...

Reply:
Well, the ellipse equation is

![](https://wikimedia.org/api/rest_v1/media/math/render/svg/863293f65080f512375d767be3a54281b1b5c2b2)

(as seen even on Wikipedia)

Reply:
`git commit -m suicide`

Reply:
Should be more readable now.

Reply:
Much better!